src/golibmc.go

```diff
@@ -148,15 +181,108 @@ use in every retrieval/storage command. default False.
 */
 func New(servers []string, noreply bool, prefix string, hashFunc int, failover bool, disableLock bool) (client *Client) {
 	client = new(Client)
-	client._imp = C.client_create()
-	runtime.SetFinalizer(client, finalizer)
+	client.servers = servers
+	client.prefix = prefix
+	client.noreply = noreply
+	client.disableLock = disableLock
+	client.hashFunc = hashFunc
+	client.failover = failover
+	client.openerCh = make(chan struct{}, connectionRequestQueueSize)
+	client.connRequest = make(map[uint64]chan *conn)
+	client.maxOpen = 1 // default value
+
+	go client.opener()
+
+	return
+}
+
+// SimpleNew to create a memcached client with default params
+func SimpleNew(servers []string) (client *Client) {
+	noreply := false
+	prefix := ""
+	hashFunc := HashCRC32
+	failover := false
+	disableLock := false
+	return New(
+		servers, noreply, prefix, hashFunc,
+		failover, disableLock,
+	)
+}
+
+func finalizer(cn *conn) {
+	C.client_destroy(cn._imp)
+}
+
+func (client *Client) lock() {
+	if !client.disableLock {
+		client.lk.Lock()
+	}
+}
+
+func (client *Client) unlock() {
+	if !client.disableLock {
+		client.lk.Unlock()
+	}
+}
+
+func (client *Client) maybeOpenNewConnections() {
+	if client.closed {
+		return
+	}
+	numRequests := len(client.connRequest)
+	if client.maxOpen > 0 {
+		numCanOpen := client.maxOpen - client.numOpen
+		if numRequests > numCanOpen {
+			numRequests = numCanOpen
+		}
+	}
+	for numRequests > 0 {
+		client.numOpen++
+		numRequests--
+		client.openerCh <- struct{}{}
+	}
+}
 
-	n := len(servers)
+func (client *Client) opener() {
+	for range client.openerCh {
+		client.openNewConnection()
+	}
+}
+
+func (client *Client) openNewConnection() {
+	cn, err := client.newConn()
+	if err != nil {
+		client.lk.Lock()
+		defer client.lk.Unlock()
+		client.numOpen--
+		client.maybeOpenNewConnections()
+		return
+	}
+	client.lk.Lock()
+	defer client.lk.Unlock()
+	if client.closed {
```

Comment:
if err == nil, need to close cn to be more precise?

Reply:
I moved this if statement before create `cn`.
https://github.com/douban/libmc/pull/57/commits/89cad5cae2136a6619ecc6b53b5c0e4dde0b876f

Reply:
i see, LGTM